controllers/resources/lease.go

```diff
@@ -0,0 +1,272 @@
+	LeaseBuffer         = time.Minute
+	RequeueIfLeaseTaken = time.Minute
+	//DefaultLeaseDuration is the default time lease would be held before it would need extending assuming escalation timeout does not exist (i.e. without escalation)
+	DefaultLeaseDuration = 10 * time.Minute
+	//max times lease would be extended - this is a conceptual variable used to calculate max time lease can be held
+	maxTimesToExtendLease = 2
+)
+
+type LeaseManager interface {
+	// ObtainNodeLease will attempt to get a node lease with the correct duration, the duration is affected by whether escalation is used and the remediation timeOut.
+	//The first return value (bool) is an indicator whether the lease was obtained, and the second return value (*time.Duration) is an indicator on when a new reconcile should be scheduled (mainly in order to extend the lease)
+	ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error)
+	//UpdateReconcileResults extends leases which needs to be extended and release those that needs to be releases, it is called at the end of reconcile and returns an updated  timeout and error to be returned by reconcile method (which are used for the next iteration of lease management)
+	UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error)
+}
+
+type nhcLeaseManager struct {
+	client             client.Client
+	commonLeaseManager lease.Manager
+	log                logr.Logger
+}
+
+func NewLeaseManager(client client.Client, log logr.Logger) (LeaseManager, error) {
+	newManager, err := lease.NewManager(client, holderIdentity)
+	if err != nil {
+		log.Error(err, "couldn't initialize lease manager")
+		return nil, err
+	}
+	return &nhcLeaseManager{
+		client:             client,
+		commonLeaseManager: newManager,
+		log:                log.WithName("nhc lease manager"),
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	return m.obtainNodeLease(remediationCR, nhc)
+}
+
+func (m *nhcLeaseManager) UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error) {
+	return m.updateReconcileResults(ctx, nhc, initialRequeue, initialErr)
+}
+
+func (m *nhcLeaseManager) obtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.calculateLeaseDurationBySingleRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
```

Comment:
I don't quite follow, maybe you can show a code snippet or share a link of what you mean ?

Reply:
nevermind, my idea won't work, we need the UUID of the node for the lease

Reply:
On the one hand I do agree it would reserve unnecessary Get call, but on the other hand I don't think it's major here since it happens when a remediation is created (which does not happen very often) and we'll need to add the node to `CreateRemediationCR` which isn't very nice IMO.
I'm slightly inclined to leave it as is but Not feeling strongly either way, so let me know what you prefer.

Reply:
wfm